osv/ecosystems/haskell.py

```diff
@@ -0,0 +1,145 @@
+Haskell ecosystem helpers.
+
+Contact the Haskell Security Response Team <security-advisories@haskell.org>
+if something is broken and you need help to fix it.
+
+"""
+
+import requests
+
+from . import config
+from .helper_base import Ecosystem, EnumerateError
+from .. import semver_index
+
+class Hackage(Ecosystem):
+  """Hackage (Haskell package index) ecosystem."""
+
+  _API_PACKAGE_URL = 'https://hackage.haskell.org/package/{package}.json'
+
+  def sort_key(self, version):
+    """Sort key.
+
+    The Haskell package version data type is defined at
+    https://hackage.haskell.org/package/Cabal-syntax/docs/Distribution-Types-Version.html
+
+    """
+    return [int(x) for x in version.split('.')]
+
+  def enumerate_versions(self,
+                         package,
+                         introduced,
+                         fixed=None,
+                         last_affected=None,
+                         limits=None):
+    """Enumerate versions."""
+    response = requests.get(
+      self._API_PACKAGE_URL.format(package=package), timeout=config.timeout)
+    if response.status_code == 404:
+      raise EnumerateError(f'Package {package} not found')
+    if response.status_code != 200:
+      raise RuntimeError(
+        f'Failed to get Hackage versions for {package} with: {response.text}'
+      )
+
+    response = response.json()
+    versions = list(resp.keys())
+
+    self.sort_versions(versions)
+    return self._get_affected_versions(versions, introduced, fixed,
+                                       last_affected, limits)
+
+
+class GHC(Ecosystem):
+  """Glasgow Haskell Compiler (GHC) ecosystem."""
+
+  _API_PACKAGE_URL = (
+    'https://gitlab.haskell.org'
+    '/api/v4/projects/3561/repository/tags?per_page=-1'
+  )
+
+  def sort_key(self, version):
+    """Sort key."""
+    return semver_index.parse(version)
+
+  @staticmethod
+  def tag_to_version(tag)
+    """Convert a tag to a release version, or return None if invalid.
+
+    GHC release tags follow the scheme:
+
+    - ghc-<major>.<minor>.<patch>-alpha<n>
+    - ghc-<major>.<minor>.<patch>-rc<n>
+    - ghc-<major>.<minor>.<patch>-release
+
+    """
+    parts = s.split('-')
+    if len(parts) == 3 and parts[0] == 'ghc':
+      if parts[2].startswith('alpha') or parts[2].startswith('rc'):
+        return '-'.join(parts[1:3])
+      elif parts[2] == 'release':
+        return parts[1]
+    return None
+
+  def enumerate_versions(self,
+                         package,
+                         introduced,
+                         fixed=None,
+                         last_affected=None,
+                         limits=None):
+    """Enumerate versions.
+
+    Different components of GHC are part of the same software release.
+    So we ignore the package (component) name.
+
+    Historical versions do not have tags in the Git repo, so we hardcode
+    the list.  See https://github.com/google/osv.dev/pull/1463 for
+    discussion.
+
+    """
+    historical_versions = [
```

Comment:
nit: maybe define the historical versions outside of `enumerate_versions()`, next to `_API_PACKAGE_URL`